docs/rpc/rpc-metrics.md

```diff
@@ -61,6 +61,8 @@ metrics can be filtered for finer grain analysis.
 >   for at least six months after it starts emitting both sets of conventions.
 > * SHOULD drop the environment variable in the next major version.
 
+Maintainers of gRPC have decided to create a different [spec](https://github.com/grpc/proposal/blob/master/A66-otel-stats.md) due to limitations on what information can be recorded by the various implementations of gRPC and differences in concepts such as a `call` and an `attempt` compared to other RPC systems.
```

Comment:
Do you think it would be possible to come up with a minimal set of generic metrics that could be used by various RPC systems/protocols? And then to add system-specific additions on top of those generic conventions?

RPC metrics are still in an experimental state, and if they don't work for gRPC I think that's a valid reason to re-think and revise them.

Reply:
hi @yashykt! we'd love to collaborate with the gRPC team to make the OpenTelemetry RPC/gRPC semantic conventions work for you

as @pyohannes mentions, the RPC/gRPC semantic conventions have not been declared stable yet, so we can still make breaking changes to them

Reply:
Sorry for the late reply. I wasn't expecting comments on this yet :)

> Do you think it would be possible to come up with a minimal set of generic metrics that could be used by various RPC systems/protocols? And then to add system-specific additions on top of those generic conventions?

I'm unsure as to whether such a minimal set of generic metrics would work for all RPC systems. Within the gRPC community itself, it was pretty hard to come up with a minimal set of metrics that worked just across the implementations of gRPC. https://github.com/grpc/proposal/blob/master/A66-otel-stats.md was the result.
I'm happy to brainstorm on a set of generic metrics here.

Reply:
Speaking as one of the gRPC TLs, I am open to discussing this further.  However, I am a little skeptical of the idea that we can define a common metric schema that would actually be usable for all RPC systems.  I think there a number of challenges here (probably not an exhaustive list):

- Different RPC systems have different architectures, and at least in gRPC, we needed a schema that reflects our architecture and may not be suitable for other RPC systems.  As mentioned in this PR, one significant difference between the gRPC schema and the that of most other RPC schemas I've seen is the difference between the `call` and `attempt` layer, reflecting the fact that gRPC has built-in retries.
- Performance is a significant focus for gRPC, and we have users that care tremendously about where the overhead is coming from, which means it's very important to have the right latency metrics.  Our current schema definition is just the start, and we expect to add more metrics in the future.  We ultimately intend to expose client-side latency measured both from the top and bottom of our stack, so that users can see how much of that delay is in the gRPC client vs. in the network.  And similarly on the server side, so that users can see how much delay is in gRPC vs. in their application.  Most existing RPC schemas that I've seen report only a single latency value and do not even specify where in the stack latency is measured from.
- While we do want to expose various metrics that are useful to gRPC users, we don't want to take on the overhead of populating unnecessary metrics, because that increases overhead and hurts performance.  As a result, it doesn't make sense to us to support both the gRPC schema and some other generic RPC schema.

As I said, I am open to discussing this further, but at the moment, it's not clear to me that there's a compelling case for a common RPC schema that can actually meet our needs here.

Reply:
Currently, semantic conventions for databases and messages are built in a way that they define metrics and attributes common to all systems. In addition to that, specific systems can add system-specific conventions (additional attributes and metrics) on top of that. Extensibility is a key feature for generic conventions.

> As I said, I am open to discussing this further, but at the moment, it's not clear to me that there's a compelling case for a common RPC schema that can actually meet our needs here.

I'm not an expert when it comes to RPC architectures, but I your proposal linked above provides a great starting port for distilling a small set of system-independent conventions. Looking at your document, I wonder if there couldn't be generic metrics like `rpc.client.call.duration` and `rpc.server.call.duration` which could be used both by gRPC and other systems. You would still be free to add additional gRPC-specific dimensions to those metrics, and you would be free to define additional gRPC-specific metrics (for example for attempts).

One main advantage of such a model is that it is much easier to adopt and support by observability vendors. As an example, many of them might utilize `rpc.client.call.duration` and `rpc.server.call.duration` for deriving service-level RED metrics, whereas fewer might undertake this task based on fractured system-specific metrics. However, by providing gRPC-specific additional dimensions and metrics, vendors and users who want to use those still can.

Reply:
My concern with that approach is that I don't think it provides an easy migration path for gRPC (or any other RPC system) to add new metrics to the common RPC metric schema over time.  If the common RPC metrics own the `rpc.` metric namespace, then I see two options for gRPC to evolve the schema:

1. gRPC adds new metrics in the `rpc.` namespace.  If we do this, then other RPC systems can do it too, which means that there could be naming conflicts that would arise as we would want to promote gRPC metrics into the common RPC metric schema, or if other RPC systems promote a metric into the common RPC metric schema and gRPC already has a metric of that name with a different definition.
2. gRPC adds new metrics in our own `grpc.` namespace.  This seems like it doesn't provide a good UX, because (a) the metrics we export would be split between two different namespaces and (b) we'd wind up having to rename these metrics as they get promoted to the common RPC metric schema.

If we use our own schema in our own metric namespace, then we don't have these problems.  This allows us to evolve the schema over time without the possibility of namespace conflicts and without breaking things for our users.

I do take your point about the advantage of a common schema for observability vendors, but right now that's not something our users have been asking for.  And my suspicion is that gRPC is popular enough that observability vendors will be willing to support our metrics, so I don't think this is going to be that big of a down-side for us.  If it turns out that I'm wrong about this, we can always add an option to support the common RPC metric schema as well as our own, but right now I'm not really seeing the benefit to us or our users.